🌑⛱️ Inductive ERModel base class

mberr opened the pull request:
This PR extracts a base class for inductive models.

This should make it easier to build your own inductive models without NodePiece representations, e.g., using textual features as in [BLP](https://github.com/dfdazac/blp).

mberr committed:
extract util

mberr committed:
add inductive ERModel

mberr committed:
upgrade inductive NP

mberr committed:
Merge remote-tracking branch 'origin/master' into inductive-ermodel

mberr committed:
rename variable

mberr committed:
do not reassign to field

mberr committed:
fix formatting

mberr committed:
update exports & resolver

mberr committed:
add init docstring

mberr committed:
fix typo & change parameter name for consistency

mberr committed:
fix passing wrong triples factory in kwargs

mberr committed:
update docstring

mberr committed:
update skip modules

mberr committed:
fix formatting

cthoyt commented on src/pykeen/models/inductive/base.py:
```diff
@@ -0,0 +1,101 @@
+        if mode in self._mode_to_representations:
+            return self._mode_to_representations[mode]
+        elif mode is None:
+            raise ValueError(f"{self.__class__.__name__} does not support inductive mode: {mode}")
```
Typo in error?

```suggestion
            raise ValueError(f"{self.__class__.__name__} does not support transductive mode")
```

cthoyt replied:
oh wait I see what's going on here. I think the check for mode being none should be first, then after check if in self._mode_to_representations, then else is saying it doesn't support the given mode

mberr replied:
The `else` branch seems to be already covered by the type checker - `Optional[InductiveMode]` is equivalent to `None | 'training' | 'validation' | 'testing'`. Thus, everything which is compliant with the type annotation ends up in one of the other two cases. This check is only useful for runtime, if a user would try to put in some other value.

So, I would suggest to get rid of one of the two cases and only differentiate between supported (i.e. a key in `_mode_to_representations`) and non-supported (everything else).

mberr replied:
https://github.com/pykeen/pykeen/pull/1106/commits/f58ade20dee8b0c42a9a00d6f88b79e2662f37d1

cthoyt replied:
Made it a bit more explicit in 2611d3a

mberr committed:
simplify logic

cthoyt committed:
Make exclusion of transductive setting explicit

If you already have to add a line for this check, might as well make more explicit exceptions

cthoyt committed:
Update base.py
